### DISCUSSION TITLE
QA about  cognifly-python's functions and features

### DISCUSSION THREAD

**Discussion by lida2003** [thecognifly/cognifly-python#1] [Q&A] [answered]
As readme suggests usages about [cognifly-python](https://github.com/thecognifly/cognifly-python)
a) Manual control
b) Python API (Pro API/School API/Streaming)

Google AIY Vision Kit (Camera + Vision Bonnet)  ---->  RPI Zero W   <--- TTL ---> FlightControl    <--- TTL --->  Matek 3901

As we know if we need school API for controlling the iNav/BF quads, there is a lot of work to do, especially pull data out of the FC and build a postion estimater.

So my question as follows:
1) All school API (which tello does), especially "forward/backward/up/down/left/right/flip/ccw/cw" is tested on RPI right?
2) API is same with  tello [Tello API](https://terra-1-g.djicdn.com/2d4dce68897a46b19fc717f3576b7c6a/Tello%20%E7%BC%96%E7%A8%8B%E7%9B%B8%E5%85%B3/For%20Tello/Tello%20SDK%20%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E-%E4%B8%AD%E6%96%871.3_1122.pdf) right ?

> Just port changes
>1. tello (8889) ==> cognify(8988)
> 2. tello (8890) ==> cognify(8989)
> 3. tello (11111) ==> cognify(8990)

3)  Where is UDP server (listens command) and UDP client (sends status back) , is it in the code of cognifly-python?
==>[cognifly-python, handles PS4/keyboard/python API](https://github.com/thecognifly/cognifly-python) 
==>[YAMSPy, handles MSP protocol throught TTL](https://github.com/thecognifly/YAMSPy)
==>[AIY-Vision-Kit-Utils, handle video analysis](https://github.com/thecognifly/AIY-Vision-Kit-Utils)

**Comment by yannbouteiller** [comment, marked as answer]
Hello,

You are right, (1) all the school API is tested on RPI, and (2) the school API largely mimics the [easytello](https://pypi.org/project/easytello/) API.

The version that is currently under development on the repo (`yb/stable` branch) works with INAV 6.

(3) the UDP logic is in the code of cognifly-python. The client/server per-se is defined [here](https://github.com/thecognifly/cognifly-python/blob/f4bcbd5e45d8844beecdf460dc3aaae54313b8ca/cognifly/utils/udp_interface.py)

**Reply by lida2003** [reply]
Really glad to see quick response. 

Further question here.
Does "forward/backward/up/down/left/right/flip/ccw/cw" tested on RPI, with proper env, has (more or less, about)[1cm  drift](https://github.com/thecognifly/cognifly-python#pro-api)?

**Reply by yannbouteiller** [reply]
It depends a lot on the conditions. If you are using the optical flow sensor over a well-textured and well-lit floor, yes. If using the optical flow sensor and the texture is bad, no, and the drone can drift badly. And if using a custom estimator with perfect position estimation and adapted PID values, there is no drift at all.

**Reply by lida2003** [reply]
Well, I'm planning to use lastest iNav (released version). And my preliminary test is based on iNav(5.0.0) + Matek 3901 module.

And now I'm going to use cognify softwares (RPI + Google AI camera) + iNav(5.0.0) + matek 3901 to setup the test. I think it should be much better than [the previous test video](https://www.bilibili.com/video/BV1Qd4y1h7UD?p=2)

What do u think?

**Reply by yannbouteiller** [reply]
The 'yb/stable' branch is not compatible with inav5, only with the yet unreleased inav6. This is because this branch uses inav6 MSP messages to fake inav-required sensors that Cognifly does not use, or even possibly all sensors (except IMU) when using a custom estimator.

The Google stuff is not used by cognifly-python when using a RPI. It may simply use a raspberry camera to stream video, that the remote can then process in Python if they wish.

**Reply by lida2003** [reply]
It really helps a lot.

a) inav6 + yb/stable  // MSP protocol compatible
b) Currently, video (AI) has nothig to do with cognifly control link  // if video, no ai required, raspberry camera is good to go.
